File: Roadmap/26 - SOLID SRP/python/Chrisdev00.py
"""
* EJERCICIO:
* Explora el "Principio SOLID de Responsabilidad Única (Single Responsibility
* Principle, SRP)" y crea un ejemplo simple donde se muestre su funcionamiento
* de forma correcta e incorrecta.
*
* DIFICULTAD EXTRA (opcional):
* Desarrolla un sistema de gestión para una biblioteca. El sistema necesita
* manejar diferentes aspectos como el registro de libros, la gestión de usuarios
* y el procesamiento de préstamos de libros.
* Requisitos:
* 1. Registrar libros: El sistema debe permitir agregar nuevos libros con 
* información básica como título, autor y número de copias disponibles.
* 2. Registrar usuarios: El sistema debe permitir agregar nuevos usuarios con
* información básica como nombre, número de identificación y correo electrónico.
* 3. Procesar préstamos de libros: El sistema debe permitir a los usuarios
* tomar prestados y devolver libros.
* Instrucciones:
* 1. Diseña una clase que no cumple el SRP: Crea una clase Library que maneje
* los tres aspectos mencionados anteriormente (registro de libros, registro de
* usuarios y procesamiento de préstamos).
* 2. Refactoriza el código: Separa las responsabilidades en diferentes clases
* siguiendo el Principio de Responsabilidad Única.
"""

# El Principio de Responsabilidad Única (Single Responsibility Principle, SRP) es uno de los cinco principios de 
# diseño de software SOLID, que fue definido por Robert C. Martin. 
# Este principio establece que una clase debe tener una, y solo una, razón para cambiar, 
# es decir, una única responsabilidad o propósito.

# Ejemplo (Forma incorrecta)

class Order:
    def __init__(self, items, tax_rate):
        self.items = items # tupla (item: precio)
        self.tax_rate = tax_rate
    
    def calculate_total(self):
        total = sum(price for item, price in self.items)
        discount = 5  # Añadir un descuento fijo
        tax = (total - discount) * self.tax_rate
        return total - discount + tax
    
    def generate_invoice(self):
        total = self.calculate_total()
        invoice = "Factura:\n"
        for item, price in self.items:
            invoice += f"{item}: ${price}\n"
        invoice += f"Total (incluyendo impuestos): ${total}\n"
        return invoice
    
    def send_confirmation_email(self, email):
        invoice = self.generate_invoice()
        print(f"Enviando email a {email} con la siguiente Factura:\n{invoice}")


order = Order([("Libro", 12.99), ("Lapicero", 0.99)], 0.1)
print(order.calculate_total())
print(order.generate_invoice())
order.send_confirmation_email("customer@example.com")


# Ejemplo (Forma correcta cumpliendo con el SRP)


class Order:
    def __init__(self, items):
        self.items = items
    
    def get_total(self):
        return sum(price for item, price in self.items)

class TaxCalculator:
    def calculate_tax(self, total, tax_rate, discount=0):
        return (total - discount) * tax_rate

class InvoiceGenerator:
    def generate_invoice(self, order, total, tax):
        invoice = "Factura:\n"
        for item, price in order.items:
            invoice += f"{item}: ${price}\n"
        invoice += f"Subtotal: ${total}\n"
        invoice += f"Impuesto: ${tax}\n"
        invoice += f"Total (incluyendo impuestos): ${total + tax}\n"
        return invoice

class EmailSender:
    def send_email(self, email, content):
        print(f"Enviando email a {email} con el siguiente contenido:\n{content}")


order = Order([("Libro", 12.99), ("Lapiz", 0.99)])
tax_calculator = TaxCalculator()
invoice_generator = InvoiceGenerator()
email_sender = EmailSender()

total = order.get_total()
tax = tax_calculator.calculate_tax(total, 0.1, discount=5)
subtotal = total - 5

factura = invoice_generator.generate_invoice(order, subtotal, tax)

email_sender.send_email("comprador@example.com", factura)


########## ------------------------------- EXTRA ---------------------------------- #######################

# Forma Incorrecta

class Library:

    def __init__(self):
        self.books = []
        self.users = []
        self.loans = []

    def add_book(self, title, author, copies):
        book = {"title": title, "author": author, "copies": copies}
        self.books.append(book)

    def add_usser(self, name, user_id, email):
        user = {"name": name, "user_id": user_id, "email": email}
        self.users.append(user)

    def borrow_book(self, user_id, title):
        for book in self.books:
            if book["title"] == title and book["copies"] > 0:
                book["copies"] -= 1
                loan = {"user_id": user_id, "title": title}
                self.loans.append(loan)
                return f"El usuario '{user_id}' se presto el libro '{title}'."
        return f"Libro '{title}' no esta disponible."
    
    def return_book(self, user_id, title):
        for loan in self.loans:
            if loan["user_id"] == user_id and loan["title"] == title:
                self.loans.remove(loan)
                for book in self.books:
                    if book["title"] == title:
                        book["copies"] += 1
                return f"Libro '{title}' retornado por el usuario '{user_id}'."
        return f"No hay registro de que el usuario '{user_id}' se haya prestado el libro '{title}'"
    

library = Library()

library.add_book("Lord of the Rings", "JRR Tolkien", 4)
library.add_book("Harry Potter", "JK Rowling", 2)

library.add_usser("Alice", "1", "alice@example.com")
library.add_usser("Matias", "2", "matias@example.com")

print(library.borrow_book("1", "Lord of the Rings"))
print(library.borrow_book("2", "Harry Potter"))
print(library.borrow_book("1", "Lord of the Rings"))
print(library.return_book("2", "The Great Gatsby"))


# Forma correcta (usando SRP)

class Library:
    def __init__(self):
        self.books = []
        self.users = []        

    def add_book(self, book):
        self.books.append(book)

    def add_user(self, user):
        self.users.append(user)

    def find_book(self, title):
        for book in self.books:
            if book.title == title:
                return book
        return None
    
    def find_user(self, user_id):
        for user in self.users:
            if user.user_id == user_id:
                return user
        return None
    
    def total_books(self):
        return sum(book.copies for book in self.books)
    
    def get_library_info(self):
        total_books = self.total_books()
        books_info = [book.show_info() for book in self.books]
        return total_books, books_info

class Book:
    def __init__(self, title, author, copies):
        self.title = title
        self.author = author
        self.copies = copies

    def show_info(self):
        return f"'{self.title}' por '{self.author}', numero de copias {self.copies}"
    
class User:
    def __init__(self, name, user_id, email):
        self.name = name
        self.user_id = user_id
        self.email = email
        self.loans = []

class LoanProcess:
    
    def borrow_book(self, library, user_id, title):
        user = library.find_user(user_id)
        book = library.find_book(title)
        if book and user:
            if book.copies > 0:
                user.loans.append(book)
                book.copies -= 1
                print(f"Libro '{book.title}' prestado a '{user.name}'")
            else:
                print(f"Este libro no esta disponible para prestamo")
        else:
            print(f"Usuario o libro no encontrado")

    def return_book(self, library, user_id, title):
        user = library.find_user(user_id)
        book = library.find_book(title)
        if book and user:
            if book in user.loans:
                user.loans.remove(book)
                book.copies += 1
                print(f"Libro '{book.title}' devuelto por '{user.name}'.")
            else:
                print(f"No hay registro de que el libro '{book.title}' haya sido prestado a '{user.name}'.")
        else:
            print("Usuario o libro no encontrado.")


library = Library()
book1 = Book("Lord of the Rings", "JRR Tolkien", 4)
book2 = Book("1984", "George Orwell", 5)
user = User("Alice", "1", "alice@example.com")
library.add_book(book1)
library.add_book(book2)
library.add_user(user)

loan_processor = LoanProcess()
loan_processor.borrow_book(library, "1", "Lord of the Rings")
loan_processor.return_book(library, "1", "Lord of the Rings")


# Total de libros en la biblioteca
total_libros, info_libros = library.get_library_info()
print(f"Total de libros en la biblioteca: {total_libros}")
print("Informacion de la biblioteca:")
for info in info_libros:
    print(info)




    

File: Roadmap/26 - SOLID SRP/javascript/Chrisdev00.js
/*
* EJERCICIO:
* Explora el "Principio SOLID de Responsabilidad Única (Single Responsibility
* Principle, SRP)" y crea un ejemplo simple donde se muestre su funcionamiento
* de forma correcta e incorrecta.
*
* DIFICULTAD EXTRA (opcional):
* Desarrolla un sistema de gestión para una biblioteca. El sistema necesita
* manejar diferentes aspectos como el registro de libros, la gestión de usuarios
* y el procesamiento de préstamos de libros.
* Requisitos:
* 1. Registrar libros: El sistema debe permitir agregar nuevos libros con 
* información básica como título, autor y número de copias disponibles.
* 2. Registrar usuarios: El sistema debe permitir agregar nuevos usuarios con
* información básica como nombre, número de identificación y correo electrónico.
* 3. Procesar préstamos de libros: El sistema debe permitir a los usuarios
* tomar prestados y devolver libros.
* Instrucciones:
* 1. Diseña una clase que no cumple el SRP: Crea una clase Library que maneje
* los tres aspectos mencionados anteriormente (registro de libros, registro de
* usuarios y procesamiento de préstamos).
* 2. Refactoriza el código: Separa las responsabilidades en diferentes clases
* siguiendo el Principio de Responsabilidad Única.
*/

// Ejemplo de Forma incorrecta 

class Order {
  constructor(items, taxRate){
    this.items = items;
    this.taxRate = taxRate;
  }

  calculateTotal() {
    let total = this.items.reduce((sum, item) => sum + item[1], 0);
    const discount = 5
    let tax = (total - discount) * this.taxRate
    return total - discount + tax;
  }

  generateInvoice(){
    const total = this.calculateTotal()
    let invoice = "Factura:\n";
    this.items.forEach(([item, price]) => {
      invoice += `${item}: $${price}\n`;
    });
    invoice += `Total (incluyendo impuestos): $${total.toFixed(2)}\n`;
    return invoice;
  }

  sendConfirmationEmail(email){
    const invoice = this.generateInvoice();
    console.log(`Enviando email a ${email} con la siguiente Factura:\n${invoice}`);

  }
}

const order1 = new Order([['item1', 10], ['item2', 20]], 0.1);
console.log(order1.generateInvoice());
order1.sendConfirmationEmail('example@example.com');


// Manera Correcta cumpliendo el Principio de Responsabilidad Única (SRP)

class Order {
  constructor(items){
    this.items = items;
  }

  getTotal() {
    return this.items.reduce((sum, item) => sum + item[1], 0);
  }
}

class TaxCalculator {
  constructor(taxRate, discount = 5) {
    this.taxRate = taxRate;
    this.discount = discount;
  }

  calculateTax(total) {
    return (total - this.discount) * this.taxRate;
  }

  calculateFinalTotal(total) {
    const tax = this.calculateTax(total);
    return total - this.discount + tax;
  }
}

class InvoiceGenerator {
  generateInvoice(order, finalTotal) {
    let invoice = "Factura:\n";
    order.items.forEach(([item, price]) => {
      invoice += `${item}: $${price}\n`;
    });
    invoice += `Total (incluyendo impuestos): $${finalTotal.toFixed(2)}\n`;
    return invoice;
  }
}

class EmailSender {
  sendConfirmationEmail(email, invoice) {
    console.log(`Enviando email a ${email} con la siguiente Factura:\n${invoice}`);
  }
}

// Uso del código
const order = new Order([['item1', 10], ['item2', 20]]);
const taxCalculator = new TaxCalculator(0.1);
const invoiceGenerator = new InvoiceGenerator();
const emailSender = new EmailSender();

const total = order.getTotal();
const finalTotal = taxCalculator.calculateFinalTotal(total);
const invoice = invoiceGenerator.generateInvoice(order, finalTotal);

console.log(invoice);
emailSender.sendConfirmationEmail('example@example.com', invoice);



/////////////-------------------------- EXTRA --------------------------- ////////////////////////////////


// Manera Incorrecta del ejercicio

class LibraryOne{

  constructor(){
    this.books = []
    this.users = []
    this.loans = []
  }

  addBook(title, author, copies){
    let book = {"title": title, "author": author, "copies": copies};
    this.books.push(book)
  }

  addUser(name, userId, email){
    let user = {"name": name, "userId": userId, "email": email};
    this.users.push(user)
  }

  borrowBook(userId, title){
    for (let book of this.books){
      if (book.title === title && book.copies > 0){
        book.copies -= 1;
        const loan = {"userId": userId, "title": title};
        this.loans.push(loan);
        return `El usuario '${userId}' se presto el libro '${title}'.`;
      }
    }
    return `Libro '${title}' no esta disponible.`
  }

  returnBook(userId, title){
    for (let loan of this.loans){
      if(loan.userId === userId && loan.title === title){
        this.loans = this.loans.filter(1 >= 1 !== loan);
        for (let book of this.books){
          if(book.title === title){
            book.copies += 1;
          }
        }
        return `Libro '${title}' retornado por el usuario '${userId}'.`;
      }
    }
    return `No hay registro de que el usuario '${userId}' se haya prestado el libro '${title}'.`;
  }
}

const library1 = new LibraryOne();

library1.addBook("Lord of the Rings", "JRR Tolkien", 4);
library1.addBook("Harry Potter", "JK Rowling", 2);

library1.addUser("Alice", "1", "alice@example.com");
library1.addUser("Matias", "2", "matias@example.com");

console.log(library1.borrowBook("1", "Lord of the Rings"));
console.log(library1.borrowBook("2", "Harry Potter"));
console.log(library1.borrowBook("1", "Lord of the Rings"));
console.log(library1.returnBook("2", "The Great Gatsby"));


// Manera Correcta cumpliendo el Principio de Responsabilidad Única (SRP)


class Library{
    constructor(){
      this.books = [];
      this.users = [];
    }
  
    addBook(book){
      this.books.push(book)
    }
  
    addUser(user){
      this.users.push(user);
    }
  
    findBook(title){
      return this.books.find(book => book.title === title) || null;
    }
  
    findUser(userId){
      return this.users.find(user => user.userId === userId) || null;
    }
  
    totalBooks(){
      return this.books.reduce((sum, book) => sum + book.copies, 0);
    }
  
    getLibraryInfo(){
      const totalBooks = this.totalBooks();
      const booksInfo = this.books.map(book => book.showInfo());
      return {totalBooks, booksInfo};
    }
  
}
  
class Book{
    constructor(title, author, copies){
      this.title = title;
      this.author = author;
      this.copies = copies;
    }
  
    showInfo(){
      return `'${this.title}' por '${this.author}', numero de copias '${this.copies}'`;
    }
}
  
class User{
    constructor(name, userId, email){
      this.name = name
      this.userId = userId
      this.email = email
      this.loans = []
    }
}
  
class LoanProcess{
    borrowBook(library, userId, title){
      let user = library.findUser(userId);
      let book = library.findBook(title);
      if(book && user){
        if(book.copies >0){
          user.loans.push(book)
          book.copies -= 1
          console.log(`Libro '${book.title}' prestado a '${user.name}'`)
        }else{
          console.log("Este libro no esta disponible para prestamo")
        }
      }else{
        console.log("Usuario o libro no encontrado")
      }
    }
  
    returnBook(library, userId, title){
      let user = library.findUser(userId);
      let book = library.findBook(title);
      if (book && user){
        const loanIndex = user.loans.indexOf(book);
        if(loanIndex !== -1){
          user.loans.splice(loanIndex, 1);
          book.copies += 1
          console.log(`Libro '${book.title}' devuelto por '${user.name}'.`)
        }else{
          console.log(`No hay registro de que el libro '${book.title}' haya sido prestado a '${user.name}'`)
        }
      }else{
        console.log("Usuario o libro no encontrado")
      }
    }
}
  
  
const library = new Library();
const book1 = new Book("Lord of the Rings", "JRR Tolkien", 4);
const book2 = new Book("1984", "George Orwell", 5);
const user = new User("Alice", "1", "alice@example.com");

library.addBook(book1);
library.addBook(book2);
library.addUser(user);

const loanProcessor = new LoanProcess();
loanProcessor.borrowBook(library, "1", "Lord of the Rings");
loanProcessor.returnBook(library, "1", "Lord of the Rings");

// Total de libros en la biblioteca
const { totalBooks, booksInfo } = library.getLibraryInfo();
console.log(`Total de libros en la biblioteca: ${totalBooks}`);
console.log("Información de la biblioteca:");
booksInfo.forEach(info => console.log(info));
